### PR TITLE
Fix and un-grey any iterator tests.

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

Stabilize the "any" tests and un-greylist them. Current approach is just to relax the distribution constraints a bit. It's an arbitrary threshold anyway, and the current algorithms are working correctly for the design, and the design is about the best we can without significant work (and probably significant slowdown).

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)

#### Related Information

- [x] There is an internal planning ticket: https://github.com/arangodb/planning/issues/4682

### Testing & Verification

This change is already covered by existing tests.

Jenkins: http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7393/

### CODE REVIEW

File: tests/js/server/shell/shell-any-nondeterministic-noncluster.js [hunk 1]
```diff
@@ -213,8 +225,16 @@ function AnySuite () {
       var statsExp = statsExpected(l, n * 200);
       var stats = statsFound(dist);
       assertEqual(stats.count, n * 200);
+      if (Math.abs(stats.average - statsExp.average)
+        >= statsExp.averageStddev * 4) {
```
Comment (jsteemann) [inline tests/js/server/shell/shell-any-nondeterministic-noncluster.js]: This is a bit inconsistent to me. Above and below we print out the values if the deviation is more than 3 times the expected value, and fail if the deviation is bigger than factor 4.
But here we print if the deviation is factor 4 and also fail at the same threshold
```suggestion
        >= statsExp.averageStddev * 3) {
```
I am not sure what the intention is for this. But if we want the thresholds to be aligned, there are a few places to adjust in this PR. The same if we want the thresholds for printing and failing to be different. :smile:

File: tests/js/server/shell/shell-any-nondeterministic-noncluster.js [hunk 2]
```diff
@@ -270,10 +298,18 @@ function AnySuite () {
       var statsExp = statsExpected(l, n * 100);
       var stats = statsFound(dist);
       assertEqual(stats.count, n * 100);
+      if(Math.abs(stats.average - statsExp.average)
```
Comment (jsteemann) [inline tests/js/server/shell/shell-any-nondeterministic-noncluster.js]:
```suggestion
      if (Math.abs(stats.average - statsExp.average)
```

**Comment by jsteemann** [comment]
Further question: IMHO this PR just increases the chances of the test passing, but it doesn't guarantee it. Is there any way we can improve on this?
IMHO the `any` algorithm is quite ok for MMFiles, but for RocksDB we want to avoid expensive seeks, so we will often return the same documents.
We could lower the test expectations to take this into account, and for example be satisfied if 100 calls to `any()` produce at least 2 different documents for RocksDB or such.

**Comment by jsteemann** [comment]
Additionally, if we are sure the test passes under almost any (pun intended) circumstances, we should rename it and remove the "nondeterministic" tag from its name.

File: tests/js/server/shell/shell-any-nondeterministic-noncluster.js [hunk 3]
```diff
@@ -167,10 +167,22 @@ function AnySuite () {
...
       var stats = statsFound(dist);
       assertEqual(stats.count, n * 100);
+      if (Math.abs(stats.average - statsExp.average)
+        >= statsExp.averageStddev * 3) {
+        require('internal').print(`real: ${stats.average}; exp: ${statsExp.average}; stddev: ${statsExp.averageStddev}}`);
+      }      
+      if (Math.abs(stats.average - statsExp.average)
+        >= statsExp.averageStddev * 3) {
+        require('internal').print(`real: ${stats.average}; exp: ${statsExp.average}; stddev: ${statsExp.averageStddev}}`);
+      }
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply tests/js/server/shell/shell-any-nondeterministic-noncluster.js]: Oh, shit, sorry, forgot to remove my debugging printouts. I'll take care of it.

[the remark this answers is not in the document]
Comment (jsteemann) [reply tests/js/server/shell/shell-any-nondeterministic-noncluster.js]: No worries

**Comment by danielhlarkin** [comment]
Interestingly enough, the RocksDB test with the old threshold managed to run about 140 times on my machine before failing once, while the MMFiles test usually failed within 10 iterations. I think our thresholds were simply too tight to account for any single run with the pseudo-random behavior.

We could potentially increase the sample size/space to tighten the variance more, in particular for MMFiles (not RocksDB since it is so slow). At that point if it fails there may be something more seriously wrong, so we could remove the non-deterministic flag. What do you think, @jsteemann?

**Comment by jsteemann** [comment]
We can definitely increase the sample size (provided it does not blow up test times) and even lower the test assumptions so that we expect the distribution to be less random than we expect now.
I am also happy with something like "if we call `any()` 1000 times we want to see at least 5 different documents or such". That is good enough, given that we cannot guarantee **any** level of randomness.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7424/

File: tests/js/server/shell/shell-any-noncluster.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        c.save({ value: i });
+        l.push(i);
+      }
+
+      var dist = getDistribution(n * 200, function () {
+        return c.any().value;
+      });
+
+      // make sure we hit everything
+      assertTrue(Object.keys(dist).length == 3);
```
Comment (jsteemann) [inline tests/js/server/shell/shell-any-noncluster.js]: jslint will complain about `==`. Instead we should use `assertEqual` here.

File: tests/js/server/shell/shell-any-noncluster.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        c.save({ value: i });
+        l.push(i);
+      }
+
+      var dist = getDistribution(n * 100, function () {
+        return c.any().value;
+      });
+
+      // make sure we hit everything
+      assertTrue(Object.keys(dist).length == 10);
```
Comment (jsteemann) [inline tests/js/server/shell/shell-any-noncluster.js]: jslint will complain about `==`. Instead we should use `assertEqual` here.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7427/

File: tests/js/server/shell/shell-any-noncluster.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    ////////////////////////////////////////////////////////////////////////////////
+    /// @brief check entropy of any(), with most documents deleted
+    ////////////////////////////////////////////////////////////////////////////////
+
+    testCheckEntropyCollectionSparse: function () {
+      var i, n, l;
+
+      n = 500;
+
+      for (i = 0; i < n; ++i) {
```
Comment (dothebart) [inline tests/js/server/shell/shell-any-noncluster.js]: please use bulk requests here. Collect documents in an array which you save with one insert after that loop.

File: tests/js/server/shell/shell-any-noncluster.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    /// @brief check entropy of any(), with many documents deleted
+    ////////////////////////////////////////////////////////////////////////////////
+
+    testCheckEntropyCollectionHalf: function () {
+      var i, n, l;
+
+      n = 500;
+
+      for (i = 0; i < n; ++i) {
+        c.save({ value: i });
```
Comment (dothebart) [inline tests/js/server/shell/shell-any-noncluster.js]: use bulk pls.

File: tests/js/server/shell/shell-any-noncluster.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    ////////////////////////////////////////////////////////////////////////////////
+
+    testCheckEntropyCollectionMore: function () {
+      var i, n, l;
+
+      n = 500;
+
+      l = [];
+      for (i = 0; i < n; ++i) {
+        c.save({ value: i });
```
Comment (dothebart) [inline tests/js/server/shell/shell-any-noncluster.js]: use bulk.

File: tests/js/server/shell/shell-any-noncluster.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    ////////////////////////////////////////////////////////////////////////////////
+
+    testCheckEntropyCollectionFew2: function () {
+      var i, n, l;
+
+      n = 10;
+
+      l = [];
+      for (i = 0; i < n; ++i) {
+        c.save({ value: i });
```
Comment (dothebart) [inline tests/js/server/shell/shell-any-noncluster.js]: use bulk

File: tests/js/server/shell/shell-any-noncluster.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    ////////////////////////////////////////////////////////////////////////////////
+
+    testCheckEntropyCollectionFew1: function () {
+      var i, n, l;
+
+      n = 3;
+
+      l = [];
+      for (i = 0; i < n; ++i) {
+        c.save({ value: i });
```
Comment (dothebart) [inline tests/js/server/shell/shell-any-noncluster.js]: use bulk

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/7457/

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/7466/

**Comment by jsteemann** [comment]
Tests blue

**Pull request merged**